src/clusterfuzz/_internal/tests/core/bot/fuzzers/centipede/centipede_engine_test.py

```diff
@@ -53,63 +70,46 @@
 ]
 
 
-def clear_output_dirs():
-  """Clears output directory."""
-  for input_dir in [CORPUS_DIR, CRASHES_DIR]:
-    if input_dir.exists():
-      shutil.rmtree(input_dir)
-    input_dir.mkdir()
-
-
-def setup_testcase(testcase):
+def setup_testcase(testcase, test_paths):
   """Sets up testcase and corpus."""
-  clear_output_dirs()
 
-  src_testcase_path = DATA_DIR / testcase
-  copied_testcase_path = CORPUS_DIR / testcase
+  src_testcase_path = test_paths.data / testcase
+  copied_testcase_path = test_paths.corpus / testcase
   shutil.copy(src_testcase_path, copied_testcase_path)
 
   return copied_testcase_path
 
 
-def setup_centipede(target_name, centipede_bin=None):
+def setup_centipede(target_name, test_paths, centipede_bin=None):
   """Sets up Centipede for fuzzing."""
   # Setup Centipede's fuzz target.
   engine_impl = engine.Engine()
-  target_path = engine_common.find_fuzzer_path(DATA_DIR, target_name)
-  sanitized_target_path = DATA_DIR / fuzzer_utils.EXTRA_BUILD_DIR / target_name
+  target_path = engine_common.find_fuzzer_path(test_paths.data, target_name)
+  sanitized_target_path = test_paths.data / fuzzer_utils.EXTRA_BUILD_DIR / target_name
 
-  # Setup Centiepde's binary.
-  if centipede_bin and centipede_bin != CENTIPEDE_BIN:
-    os.rename(CENTIPEDE_BIN, CENTIPEDE_BIN_TMP)
-    os.rename(centipede_bin, CENTIPEDE_BIN)
+  # Setup Centipede's binary.
+  if centipede_bin and centipede_bin != test_paths.centipede:
+    os.rename(centipede_bin, test_paths.centipede)
 
   return engine_impl, target_path, sanitized_target_path
 
 
-def reset_centipede(centipede_bin=None):
-  """Sets up Centipede for fuzzing."""
-  # Reset Centiepde's binary.
-  if centipede_bin and centipede_bin != CENTIPEDE_BIN:
-    os.rename(CENTIPEDE_BIN, centipede_bin)
-    os.rename(CENTIPEDE_BIN_TMP, CENTIPEDE_BIN)
-
-
 @test_utils.integration
 class IntegrationTest(unittest.TestCase):
   """Integration tests."""
 
-  def setUp(self):
-    self.maxDiff = None  # pylint: disable=invalid-name
+  def run(self, *args, **kwargs):
     test_helpers.patch_environ(self)
+    with get_test_paths() as test_paths:
+      self.test_paths = test_paths
+      os.environ['BUILD_DIR'] = str(self.test_paths.data)
+      super().run(*args, **kwargs)
```

Comment:
Not sure if I follow here:
Do we move the preparation steps in `run()` to save time?
Thanks!

Reply:
No. We do it so that the testcases are run within the ```with_gets_paths() as test_paths:``` block. We can't do that with setup. We don't 100% need to do this, but this trick saved us a ton of repetition.